src/dma/mod.rs

```diff
@@ -0,0 +1,799 @@
+                // there's nothing we can do apart from panicking.  TODO: Is
+                // this the best solution ? The closure based approach seems
+                // necessary if we want to support BBqueue.
+                panic!("Overrun");
+            }
+
+            if current_buffer == CurrentBuffer::DoubleBuffer {
+                self.stream.set_memory_address(new_buf_ptr as u32);
+
+                // Check again if an overrun occurred, the buffer address won't
+                // be updated in that case
+                if self.stream.get_memory_address() != new_buf_ptr as u32 {
+                    panic!("Overrun");
+                }
+
+                // "Subsequent reads and writes cannot be moved ahead of
+                // preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                self.buf.replace(new_buf);
+                return Ok(r.1);
+            } else {
+                self.stream
+                    .set_memory_double_buffer_address(new_buf_ptr as u32);
+                if self.stream.get_memory_double_buffer_address()
+                    != new_buf_ptr as u32
+                {
+                    panic!("Overrun");
+                }
+
+                // "Subsequent reads and writes cannot be moved ahead of
+                // preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                self.double_buf.replace(new_buf);
+                return Ok(r.1);
+            }
+        }
+        self.stream.disable();
+        self.stream.clear_transfer_complete_interrupt();
+
+        // "No re-ordering of reads and writes across this point is allowed"
+        compiler_fence(Ordering::SeqCst);
+
+        // Can never fail, we never let the Transfer without a buffer
+        let old_buf = self.buf.take().unwrap();
+        let r = f(old_buf, CurrentBuffer::FirstBuffer);
+        let mut new_buf = r.0;
+
+        let (buf_ptr, buf_len) = new_buf.write_buffer();
+        self.stream.set_memory_address(buf_ptr as u32);
+        self.stream.set_number_of_transfers(buf_len as u16);
+        self.buf.replace(new_buf);
+
+        // Ensure that all transfers to normal memory complete before
+        // subsequent memory transfers.
+        //
+        // The new memory buffer is almost certainly in normal memory, so we
+        // ensure that all transfers there complete before proceeding to enable
+        // the stream
+        cortex_m::asm::dmb();
+
+        // "Preceding reads and writes cannot be moved past subsequent writes"
+        compiler_fence(Ordering::Release);
+
+        self.stream.enable();
+
+        Ok(r.1)
+    }
+}
+
+impl<STREAM, PERIPHERAL, DIR, BUF> Drop
+    for Transfer<STREAM, PERIPHERAL, DIR, BUF>
+where
+    STREAM: Stream,
+    PERIPHERAL: TargetAddress<DIR>,
+    DIR: Direction,
+    BUF: WriteBuffer<Word = <PERIPHERAL as TargetAddress<DIR>>::MemSize>
+        + 'static,
+{
+    fn drop(&mut self) {
```

Comment:
An example is using a circular transfer to write a single word into the SPI TXFIFO to trigger a transaction. The transfer may be triggered by a timer overflow, which results in a periodic SPI read. The actual transfer and TX contents are don't care, but the transfer should continue on infinitely and there's no value of keeping the `Transfer` around since it is never modified.

Reply:
I've proposed one way of addressing this in https://github.com/stm32-rs/stm32h7xx-hal/pull/178